agent/connect/ca/provider_vault.go

```diff
@@ -637,3 +695,64 @@ func ParseVaultCAConfig(raw map[string]interface{}) (*structs.VaultCAProviderCon
 
 	return &config, nil
 }
+
+func vaultLogin(client *vaultapi.Client, authMethod *structs.VaultAuthMethod) (*vaultapi.Secret, error) {
+	loginPath, err := configureVaultAuthMethod(authMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.Logical().Write(loginPath, authMethod.Params)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil || resp.Auth == nil || resp.Auth.ClientToken == "" {
+		return nil, fmt.Errorf("login response did not return client token")
+	}
+
+	return resp, nil
+}
+
+func configureVaultAuthMethod(authMethod *structs.VaultAuthMethod) (loginPath string, err error) {
+	if authMethod.MountPath == "" {
+		authMethod.MountPath = authMethod.Type
+	}
+
+	switch authMethod.Type {
+	case VaultAuthMethodTypeKubernetes:
+		// For the Kubernetes Auth method, we will try to read the JWT token
+		// from the default service account file location if jwt was not provided.
+		if jwt, ok := authMethod.Params["jwt"]; !ok || jwt == "" {
+			serviceAccountToken, err := os.ReadFile(defaultK8SServiceAccountTokenPath)
+			if err != nil {
+				return "", err
+			}
+
+			authMethod.Params["jwt"] = string(serviceAccountToken)
+		}
+		loginPath = fmt.Sprintf("auth/%s/login", authMethod.MountPath)
+	// These auth methods require a username for the login API path.
+	case VaultAuthMethodTypeLDAP, VaultAuthMethodTypeUserpass, VaultAuthMethodTypeOkta, VaultAuthMethodTypeRadius:
+		// Get username from the params.
+		if username, ok := authMethod.Params["username"]; ok {
+			loginPath = fmt.Sprintf("auth/%s/login/%s", authMethod.MountPath, username)
+		} else {
+			return "", fmt.Errorf("failed to get 'username' from auth method params")
+		}
+	// This auth method requires a role for the login API path.
+	case VaultAuthMethodTypeOCI:
+		if role, ok := authMethod.Params["role"]; ok {
+			loginPath = fmt.Sprintf("auth/%s/login/%s", authMethod.MountPath, role)
+		} else {
+			return "", fmt.Errorf("failed to get 'role' from auth method params")
+		}
+	case VaultAuthMethodTypeToken:
+		return "", fmt.Errorf("'token' auth method is not supported via auth method configuration; " +
+			"please provide the token with the 'token' parameter in the CA configuration")
+	// The rest of the auth methods use auth/<auth method path> login API path.
+	default:
+		loginPath = fmt.Sprintf("auth/%s/login", authMethod.MountPath)
+	}
```

Comment:
this may just me being confused here, but if the `authMethod.Type` was `ldap12345` (something not in our constants), we hit the default here, right?

In that case, we will err out down the code execution path when we call the vault api and that should still be caught in some err handling 🤔 

-- 

Should we, instead, define all the supported `VaultAuthMethod*`s in their own case and err out on `default` as an unrecognized option?

Reply:
>his may just me being confused here, but if the authMethod.Type was ldap12345 (something not in our constants), we hit the default here, right?

right. Yep we will rely on vault there.

>Should we, instead, define all the supported VaultAuthMethod*s in their own case and err out on default as an unrecognized option?

I was on the fence on whether we should do that or if we should instead allow for some future auth method to be supported. This will only be the case if the auth method is using the `auth/<auth method name>/login` path though. In the end, I decided to keep it open for that future case to decrease the maintenance burden here a bit but definitely open to changing it like you suggest. WDYT?

Reply:
Thanks for the context here! 

> WDYT?

Hm, I have a slight preference towards doing *some validation* somewhere. If that make sense.

So if we treated `default` as an error case, it'd also act a bit as validation in my mind (that the auth method provided is something we can deal with).

That said, this is not a stick point on my end right now! So if you wanted to add any validation here or a layer above, I can put some 👀 on it!

Reply:
That makes sense! I've changed it to error out if the auth method is not supported and added a list of all currently supported methods. I've also added a test for this function to verify that it does the right thing. 

Let me know if that's what you had in mind @FFMMM!